src/nvim/event/rstream.c

```diff
@@ -167,9 +169,19 @@ static void fread_idle_cb(uv_idle_t *handle)
   invoke_read_cb(stream, false);
 }
 
-static void invoke_read_cb(Stream *stream, bool eof)
+static void read_event(void **argv)
 {
+  Stream *stream = argv[0];
+  bool eof = (uintptr_t)argv[1];
   if (stream->read_cb) {
     stream->read_cb(stream, stream->buffer, stream->data, eof);
   }
+  if (eof && stream->internal_eof_cb) {
+    stream->internal_eof_cb(stream, stream->internal_data);
+  }
+}
+
+static void invoke_read_cb(Stream *stream, bool eof)
+{
+  queue_put(stream->events, read_event, 2, stream, (void *)(uintptr_t)eof);
```

Comment:
Makes sense.

But this would only be a problem when stdout and stderr both represent the same file descriptor right? If that's the case, then only pty jobs can cause this kind of problem and it could be fixed by setting up a shared RBuffer instance for both stdout and stderr on pty_process.c

Is there any other scenario where you can see this reordering would cause a problem?